### PR TITLE
add never_triggered

### PR DESCRIPTION
**Pull request by Mariusthvdb**
Add never_triggered to the last-triggered secondary_info.
as it stands, the secondary_info line is empty when used on an automation or script that hasn't been triggered yet, while the user expects info on that line.

my suggestion would be to add:
```
"automation": {
        "last_triggered": "Last triggered",
        "never_triggered":"Never triggered",
        "trigger": "Execute"
      }
```
and
```
"script": {
        "last_triggered": "Last triggered",
        "never_triggered":"Never triggered",
        "execute": "Execute"
      }
```
to the [translation file](https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/translations/en.json).

### CODE REVIEW

File: src/panels/lovelace/components/hui-generic-entity-row.ts [hunk 1]
```diff
@@ -105,14 +105,15 @@ class HuiGenericEntityRow extends LitElement {
...
-                `
+              : this.config.secondary_info === "last-triggered"
+               ? stateObj.attributes.last_triggered
+               ? html`
+                   <ha-relative-time
+                     .hass=${this.hass}
+                     .datetime=${stateObj.attributes.last_triggered}
+                   ></ha-relative-time>
+                 `
+               : this.hass.localize("ui.panel.lovelace.entities.never_triggered")
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]: You should add this to https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/translations/en.json

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: ok I will, referring to this PR.

Comment (bramkragten) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: No, in this PR

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: see https://github.com/home-assistant/home-assistant-polymer/pull/4444

**Review by Mariusthvdb — COMMENTED** [review]
thanks for your feedback, appreciated!

File: src/panels/lovelace/components/hui-generic-entity-row.ts (hunk 1 above)
Comment (iantrich) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: It should be in this PR, not a separate one

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: ok, I will, if I can find how to do that... how can I add a second file to this PR to edit?

Comment (bramkragten) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: Just edit it in the same branch: https://github.com/Mariusthvdb/home-assistant-polymer/tree/patch-3

You would also need to fix the lint issues (running prettier would fix that for you)

Comment (iantrich) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]: this path doesn't match what you added to en.json

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: ok so now I am confused. Please bare with me..
I believed you suggested this path, and so I added that to the pr of the hui-generic-entity-row.ts

following that Bram asked me to add to the translation file https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/translations/en.json

which of both should I change?

Comment (bramkragten) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: It is the right file it is just that `ui.panel.lovelace.entities.never_triggered` isn't what you created that is: `ui.card.automation.never_triggered`

Comment (bramkragten) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: And you should not add keys that we don't use in the code, like the other keys you added

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: I see, I searched for automation and script, but should have looked for path lovelace is it? I think this is the only hat to lovelace I can find, is that what you refer to? https://github.com/home-assistant/home-assistant-polymer/blob/4e5406b27b0a14788e80859a96d2b54871954b39/src/translations/en.json#L1575 ?

if so, there is no path to entities yet? I should add this:
"entitities": {"never_triggered":"Never triggered"},

Comment (bramkragten) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: Put it in here:
https://github.com/home-assistant/home-assistant-polymer/blob/4e5406b27b0a14788e80859a96d2b54871954b39/src/translations/en.json#L1700

File: src/translations/en.json [hunk 2]
```diff
@@ -458,6 +459,8 @@
         "activate": "Activate"
       },
       "script": {
+        "last_triggered": "Last triggered",
+        "never_triggered":"Never triggered",
```
Comment (iantrich) [inline src/translations/en.json]: Remove these two. You are not using them.

File: src/panels/lovelace/components/hui-generic-entity-row.ts [hunk 3]
```diff
@@ -105,14 +105,17 @@ class HuiGenericEntityRow extends LitElement {
...
+              : this.config.secondary_info === "last-triggered"
+              ? stateObj.attributes.last_triggered
+                ? html`
+                    <ha-relative-time
+                      .hass=${this.hass}
+                      .datetime=${stateObj.attributes.last_triggered}
+                    ></ha-relative-time>
+                  `
+                : this.hass.localize(
+                    "ui.panel.lovelace.entities.never_triggered"
```
Comment (iantrich) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]: The path you pass is still not the same as what you out in en.json

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: Ive just corrected that, deleted the incorrect additions and added never_triggered to line 1700 in en.json, following Brams instruction

File: src/translations/en.json [hunk 4]
```diff
@@ -1698,6 +1698,7 @@
               "optional": "Optional"
             },
             "entities": {
+              "never_triggered":"Never triggered",
```
Comment (bramkragten) [inline src/translations/en.json]:
```suggestion
              "never_triggered": "Never triggered",
```

File: src/panels/lovelace/components/hui-generic-entity-row.ts (hunk 3 above)
Comment (bramkragten) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]:
```suggestion
                    "ui.panel.lovelace.cards.entities.never_triggered"
```

File: src/panels/lovelace/components/hui-generic-entity-row.ts [hunk 5]
```diff
@@ -105,14 +105,17 @@ class HuiGenericEntityRow extends LitElement {
...
+              : this.config.secondary_info === "last-triggered"
+              ? stateObj.attributes.last_triggered
+                ? html`
+                    <ha-relative-time
+                      .hass=${this.hass}
+                      .datetime=${stateObj.attributes.last_triggered}
+                    ></ha-relative-time>
+                  `
+                : this.hass.localize(
+                    "ui.panel.lovelace.cards.entities.never_triggered"
```
Comment (bramkragten) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]:
```suggestion
                    "ui.panel.lovelace.card.entities.never_triggered"
```

**Comment by bramkragten** [comment]
Please also test your PR

**Comment by Mariusthvdb** [comment]
> Please also test your PR

ok let me know how?

**Comment by bramkragten** [comment]
Follow there instructions: https://developers.home-assistant.io/docs/en/frontend_development.html

**Comment by Mariusthvdb** [comment]
That is a bit overtaxing at the moment I fear... setting up a full development environment I wouldn't know where to start with, for such a small edit..
Hope to have a spare device for that anytime soon, please have patience.

File: src/panels/lovelace/components/hui-generic-entity-row.ts [hunk 6]
```diff
@@ -105,14 +105,17 @@ class HuiGenericEntityRow extends LitElement {
...
+                ? html`
+                    <ha-relative-time
+                      .hass=${this.hass}
+                      .datetime=${stateObj.attributes.last_triggered}
+                    ></ha-relative-time>
+                  `
+                : this.hass.localize(
+                    "ui.panel.lovelace.card.entities.never_triggered"
+                  )
               : ""}
```
Comment (ludeeus) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]:
```suggestion
              }
```

Comment (Mariusthvdb) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: Hi Ludeeus,

This was already in the original file. Should it be taken out ?

Comment (ludeeus) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: No, just looked a little off with all the nesting.

**Comment by Mariusthvdb** [comment]
@bramkragten 
having some difficulties setting up my dev environment, which won't go beyond setting up an integration (Sonos) and doesn't finish the full startup sequence... because of this, I can't set up any testing yet I fear. Is there any way you guys can test if this works now, so we can proceed with the PR?

**Comment by ludeeus** [comment]
Tested, it does not work, probably needs a check against `null` and/or `undefined`

**Comment by Mariusthvdb** [comment]
a check is what I had in my first suggestion, which was later on superseded with Ian 's code. 

what would be best? to first check if not null? like
```
              : this.config.secondary_info === "last-triggered" &&
                 stateObj.attributes.last_triggered !== null
                 ? html`
                     <ha-relative-time
                       .hass=${this.hass}
                       .datetime=${stateObj.attributes.last_triggered}
                     ></ha-relative-time>
                   `
                 : this.hass.localize(
                     "ui.panel.lovelace.card.entities.never_triggered"
                   )
```

**Comment by bramkragten** [comment]
Probably the second with a falsy check instead of `null` (as you now don't catch undefined)

**Comment by Mariusthvdb** [comment]
since this is new to me, I'm not sure what I should add. Checked all never_triggered automations and scripts in HA states, and in template editor. 

In states they all `show last_triggered: null`, while in templates attributes_last_triggered == None or none. Never unknown or null.

Schermafbeelding 2020-01-13 om 13 40 49

Schermafbeelding 2020-01-13 om 13 42 32

If that still leaves you needing an extra falsy check, please let me know what that would need to be besides testing for `!== null`

**Comment by iantrich** [comment]
The problem is not the check it is your translation path. You added your translation to ui.panel.lovelace.editor.entities.never_triggered in `en.json`

**Review by iantrich — APPROVED** [review]

**Comment by Mariusthvdb** [comment]
wow, thanks Ian, really appreciated!
was seriously confused by all of this, yet so simple....
glad it is finally approved!
